### PR TITLE
User can select who goes first

### PR DESCRIPTION
**Pull request by lpenzey**
User is now prompted to select which player goes first. They must enter
the exact player/opponent token to proceed, otherwise they're shown an
error message and are prompted again to choose which player they would
like to go first. Their choice is then set in the current_player field
in the State struct.

The functionality is implemented in Game.play, where the turn is
determined by who the current player is as determined by querying
state.current_player.

### CODE REVIEW

File: lib/TTT/validation.ex [hunk 1]
```diff
@@ -39,6 +39,23 @@ defmodule Validation do
     end
   end
 
+  def choose_first_player(io, player_token, opponent_token) do
+    io.retrieve(
+      Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
+    )
+    |> clean()
+    |> validate_first_player(io, player_token, opponent_token)
```
Comment (ssunday) [inline lib/TTT/validation.ex]: Would it make more sense to throw an error in validate, then have this case off of it and recurse within itself?

Comment (lpenzey) [reply lib/TTT/validation.ex]: like this?

``` 
  def choose_first_player(io, player_token, opponent_token) do
    first_player =
      io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()

    case validate_first_player(first_player, player_token, opponent_token) do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```

Comment (ssunday) [reply lib/TTT/validation.ex]: Yeah, or piping validate_first_player to `|> clean()`

```
def choose_first_player(io, player_token, opponent_token) do
    io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()
      |> validate_first_player(player_token, opponent_token)
      |> case do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```

**Comment by rmulhol** [comment]
How's the performance if the game is setup with the unbeatable computer player going first?

File: lib/TTT/game.ex [hunk 2]
```diff
@@ -5,32 +5,40 @@ defmodule Game do
...
-    human_move(state, io)
-    |> opponent(io, mode)
-    |> check_status(io, mode)
+    case state.current_player do
+      ^player ->
+        human_move(state, io, mode)
+        |> check_status(io, mode)
+
+      ^opponent ->
+        opponent(state, io, mode)
```
Comment (rmulhol) [inline lib/TTT/game.ex:39]: This is pretty minor but I'd advise against overloading an overloaded function with a variable of the same name

**Comment by lpenzey** [comment]
> How's the performance if the game is setup with the unbeatable computer player going first?

Quite slow (~24 seconds). Easy fix is to hard-code the first move to be in one of the corners, but I was trying to tackle it using processes to speed things up.

**Pull request merged**